tests/devices/unit_tests/test_zocalo_results.py

```diff
@@ -250,3 +260,183 @@ async def test_when_exception_caused_by_zocalo_message_then_exception_propagated
+    )
+
+
+@patch("dodal.devices.zocalo.zocalo_results.LOGGER")
+@patch("dodal.devices.zocalo.zocalo_results._get_zocalo_connection", autospec=True)
+async def test_if_zocalo_results_timeout_from_one_source_then_warn(
+    mock_connection, mock_logger, RE: RunEngine
+):
+    zocalo_results = ZocaloResults(
+        name="zocalo", zocalo_environment="dev_artemis", use_cpu_and_gpu_zocalo=True
+    )
+    await zocalo_results.connect()
+    await zocalo_results.stage()
+    zocalo_results._raw_results_received.get = MagicMock(
+        side_effect=[{"ispyb_ids": {"test": 0}, "results": []}, Empty]
+    )
+    RE(bps.trigger(zocalo_results, wait=False))
+    mock_logger.warning.assert_called_with(
+        "Zocalo results from GPU timed out. Using results from CPU"
+    )
+
+
+@patch("dodal.devices.zocalo.zocalo_results.LOGGER")
+@patch("dodal.devices.zocalo.zocalo_results._get_zocalo_connection", autospec=True)
+async def test_if_cpu_results_arrive_before_gpu_then_warn(
+    mock_connection, mock_logger, RE: RunEngine
+):
+    zocalo_results = ZocaloResults(
+        name="zocalo", zocalo_environment="dev_artemis", use_cpu_and_gpu_zocalo=True
+    )
+    await zocalo_results.connect()
+    await zocalo_results.stage()
+    zocalo_results._raw_results_received.get = MagicMock(
+        return_value={"ispyb_ids": {"test": 0}, "results": []}
+    )
+    RE(bps.trigger(zocalo_results, wait=False))
+    mock_logger.warning.assert_called_with(
+        "Recieved zocalo results from CPU before GPU"
+    )
+
+
+@patch("dodal.devices.zocalo.zocalo_results.LOGGER")
+@patch("dodal.devices.zocalo.zocalo_results._get_zocalo_connection", autospec=True)
+async def test_warning_if_results_are_different(
+    mock_connection, mock_logger, RE: RunEngine
+):
+    zocalo_results = ZocaloResults(
+        name="zocalo", zocalo_environment="dev_artemis", use_cpu_and_gpu_zocalo=True
+    )
+    await zocalo_results.connect()
+    await zocalo_results.stage()
+    zocalo_results._raw_results_received.get = MagicMock(
+        side_effect=[
+            {"ispyb_ids": {}, "results": [{"test": 0}]},
+            {"ispyb_ids": {}, "results": [{"test": 1}]},
+        ]
+    )
+    RE(bps.trigger(zocalo_results, wait=False))
+    mock_logger.warning.assert_called_with(
+        "Results differed in test: CPU contains 0 while GPU contains 1 \n"
+    )
+
+
+@patch("dodal.devices.zocalo.zocalo_results._get_zocalo_connection", autospec=True)
+async def test_if_zocalo_results_timeout_before_any_results_then_error(
+    mock_connection, RE: RunEngine
+):
+    zocalo_results = ZocaloResults(
+        name="zocalo", zocalo_environment="dev_artemis", use_cpu_and_gpu_zocalo=True
+    )
+    await zocalo_results.connect()
+    await zocalo_results.stage()
+    zocalo_results._raw_results_received.get = MagicMock(side_effect=Empty)
+    with pytest.raises(NoResultsFromZocalo):
+        await zocalo_results.trigger()
```

Comment:
There was a strange thing here where this test would fail if I used `RE(bps.trigger(zocalo_results)` instead, even though debugging the pytest showed the exception was being raised